File: packages/schemas/src/quest.ts
import { QuestStatusEnum, QuestTypeEnum } from "@planner/enums/quest";
import z from "zod";
import { BaseEntitySchema } from "./common";

export const QuestRewardColonySchema = z.object({
	buildingMaterials: z.number().int().min(0).optional(),
	colonistCount: z.number().int().min(0).optional(),
	food: z.number().int().min(0).optional(),
	gold: z.number().int().min(0).optional(),
	morale: z.number().int().min(0).max(100).optional(),
});

export const QuestRewardLootItemSchema = z.object({
	description: z.string().optional(),
	name: z.string(),
	quantity: z.number().int().min(1).optional(),
});

export const QuestRewardSchema = z.object({
	colony: QuestRewardColonySchema.optional(),
	loot: z.array(QuestRewardLootItemSchema).optional(),
});

export const QuestSchema = BaseEntitySchema.extend({
	campaignId: z.uuid(),
	completedAt: z.date().nullable().optional(),
	deletedAt: z.date().nullable().optional(),
	description: z.string().nullable().optional(),
	questGiverId: z.uuid().nullable().optional(),
	reward: QuestRewardSchema.nullable().optional(),
	status: z.enum(QuestStatusEnum),
	title: z.string(),
	type: z.enum(QuestTypeEnum).optional(),
});

export const CreateQuestRequestSchema = z.object({
	campaignId: z.uuid(),
	description: z.string().optional(),
	questGiverId: z.uuid().optional(),
	reward: QuestRewardSchema.optional(),
	status: z.enum(QuestStatusEnum),
	title: z.string(),
	type: z.enum(QuestTypeEnum).optional(),
});

export const CreateQuestResponseSchema = z.object({ quest: QuestSchema });
export const GetQuestRequestSchema = z.object({ id: z.uuid() });
export const GetQuestResponseSchema = z.object({ quest: QuestSchema });
export const ListQuestsByCampaignRequestSchema = z.object({
	campaignId: z.uuid(),
});
export const ListQuestsByCampaignResponseSchema = z.object({
	quests: z.array(QuestSchema),
});

export const UpdateQuestRequestSchema = z.object({
	description: z.string().optional(),
	id: z.uuid(),
	reward: QuestRewardSchema.optional(),
	status: z.enum(QuestStatusEnum).optional(),
	title: z.string().optional(),
	type: z.enum(QuestTypeEnum).optional(),
});

export const UpdateQuestResponseSchema = z.object({ quest: QuestSchema });

export const RemoveQuestRequestSchema = z.object({ id: z.uuid() });

export const RemoveQuestResponseSchema = z.object({});

export type QuestRewardColony = z.infer<typeof QuestRewardColonySchema>;
export type QuestRewardLootItem = z.infer<typeof QuestRewardLootItemSchema>;
export type QuestReward = z.infer<typeof QuestRewardSchema>;
export type Quest = z.infer<typeof QuestSchema>;
export type CreateQuestRequest = z.infer<typeof CreateQuestRequestSchema>;
export type UpdateQuestRequest = z.infer<typeof UpdateQuestRequestSchema>;
export type RemoveQuestRequest = z.infer<typeof RemoveQuestRequestSchema>;

File: apps/api/internal/models/quest.go
package model

import (
	"database/sql"
	"time"
)

type QuestStatus string

const (
	QuestStatusActive      QuestStatus = "ACTIVE"
	QuestStatusCompleted   QuestStatus = "COMPLETED"
	QuestStatusFailed      QuestStatus = "FAILED"
	QuestStatusUnspecified QuestStatus = "UNSPECIFIED"
)

type QuestType string

const (
	QuestTypeMainland QuestType = "MAINLAND"
	QuestTypeColony   QuestType = "COLONY"
)

type QuestRewardColony struct {
	Gold              *int32 `json:"gold,omitempty"`
	Food              *int32 `json:"food,omitempty"`
	BuildingMaterials *int32 `json:"buildingMaterials,omitempty"`
	ColonistCount     *int32 `json:"colonistCount,omitempty"`
	Morale            *int32 `json:"morale,omitempty"`
}

type QuestRewardLootItem struct {
	Name        string  `json:"name"`
	Quantity    *int32  `json:"quantity,omitempty"`
	Description *string `json:"description,omitempty"`
}

type QuestReward struct {
	Colony *QuestRewardColony  `json:"colony,omitempty"`
	Loot   []QuestRewardLootItem `json:"loot,omitempty"`
}

type Quest struct {
	ID           string
	CampaignID   string
	Title        string
	Status       QuestStatus
	Description  sql.NullString
	QuestGiverID sql.NullString
	Reward       *QuestReward
	CompletedAt  sql.NullTime
	DeletedAt    sql.NullTime
	CreatedAt    time.Time
	UpdatedAt    time.Time
	Type         *QuestType
}

type CreateQuestRequest struct {
	CampaignID   string
	Title        string
	Status       QuestStatus
	Description  sql.NullString
	QuestGiverID sql.NullString
	Reward       *QuestReward
	Type         *QuestType
}

type UpdateQuestRequest struct {
	ID          string
	Title       *string
	Status      *QuestStatus
	Description sql.NullString
	CampaignID  string
	Type        *QuestType
	Reward      *QuestReward
}
